File: app/src/main/java/com/aaronhalbert/nosurfforreddit/repository/AuthenticatorUtils.java
package com.aaronhalbert.nosurfforreddit.repository;

import android.content.Intent;
import android.net.Uri;

import com.aaronhalbert.nosurfforreddit.BuildConfig;
import com.aaronhalbert.nosurfforreddit.exceptions.NoSurfAccessDeniedLoginException;
import com.aaronhalbert.nosurfforreddit.exceptions.NoSurfLoginException;

import java.util.UUID;

public class AuthenticatorUtils {
    private static final String RESPONSE_TYPE = "code";
    private static final String DURATION = "permanent";
    private static final String SCOPE = "identity mysubreddits read";
    private static final String AUTH_URL_RESPONSE_TYPE = "&response_type=";
    private static final String AUTH_URL_STATE = "&state=";
    private static final String AUTH_URL_REDIRECT_URI = "&redirect_uri=";
    private static final String AUTH_URL_DURATION = "&duration=";
    private static final String AUTH_URL_SCOPE = "&scope=";

    public static String buildAuthUrl() {
        return BuildConfig.AUTH_URL_BASE
                + BuildConfig.CLIENT_ID
                + AUTH_URL_RESPONSE_TYPE
                + RESPONSE_TYPE
                + AUTH_URL_STATE
                + UUID.randomUUID().toString()
                + AUTH_URL_REDIRECT_URI
                + BuildConfig.REDIRECT_URI
                + AUTH_URL_DURATION
                + DURATION
                + AUTH_URL_SCOPE
                + SCOPE;
    }

    /* get the one-time use code returned by the Reddit auth server. We later exchange it for a
     * bearer token */
    public static String extractCodeFromIntent(Intent i) throws NoSurfLoginException {
        final String ERROR = "error";
        final String CODE = "code";
        final String ACCESS_DENIED_ERROR_CODE = "access_denied";

        Uri uri;
        String error;
        String code;

        if (i.getData() != null) {
            uri = i.getData();
        } else {
            throw new NoSurfLoginException();
        }

        if (uri != null) {
            error = uri.getQueryParameter(ERROR);
            code = uri.getQueryParameter(CODE);
        } else {
            throw new NoSurfLoginException();
        }

        if (ACCESS_DENIED_ERROR_CODE.equals(error)) {
            throw new NoSurfAccessDeniedLoginException();
        } else if ("".equals(error)) {
            throw new NoSurfLoginException();
        }

        if (code != null && !"".equals(code)) {
            return code;
        } else {
            throw new NoSurfLoginException();
        }
    }
}

File: app/src/main/java/com/aaronhalbert/nosurfforreddit/repository/AuthenticatorUtils.kt
package com.aaronhalbert.nosurfforreddit.repository

import android.content.Intent
import com.aaronhalbert.nosurfforreddit.BuildConfig
import okhttp3.HttpUrl
import java.util.*

const val CLIENT_ID_PARAMETER_NAME = "client_id"
const val AUTH_URL_RESPONSE_TYPE = "response_type"
const val RESPONSE_TYPE = "code"
const val AUTH_URL_STATE = "state"
const val AUTH_URL_REDIRECT_URI = "redirect_uri"
const val AUTH_URL_DURATION = "duration"
const val DURATION = "permanent"
const val AUTH_URL_SCOPE = "scope"
const val SCOPE = "identity mysubreddits read"
const val ERROR = "error"
const val CODE = "code"

class AuthenticatorUtils {

    fun buildAuthUrl(): String {
        return HttpUrl
                .parse(BuildConfig.AUTH_URL_BASE)
                ?.newBuilder()
                ?.addQueryParameter(CLIENT_ID_PARAMETER_NAME, BuildConfig.CLIENT_ID)
                ?.addQueryParameter(AUTH_URL_RESPONSE_TYPE, RESPONSE_TYPE)
                ?.addQueryParameter(AUTH_URL_STATE, UUID.randomUUID().toString())
                ?.addQueryParameter(AUTH_URL_REDIRECT_URI, BuildConfig.REDIRECT_URI)
                ?.addQueryParameter(AUTH_URL_DURATION, DURATION)
                ?.addQueryParameter(AUTH_URL_SCOPE, SCOPE)
                ?.build()
                .toString()
    }

    /* Get the one-time use code returned by the Reddit auth server. We later exchange it for a
     * bearer token.
     *
     * Returning an empty string indicates failure; a non-empty value indicates success.
     *
     * Although Reddit sends a descriptive string in case of an error, this function does not
     * report it; all errors are treated as failure and reported as an empty code string
     * regardless of their type. */
    fun extractCodeFromIntent(intent: Intent): String {
        val uri = intent.data ?: return ""
        val error = uri.getQueryParameter(ERROR) ?: ""
        val code = uri.getQueryParameter(CODE) ?: return ""

        when {
            ("" != error) -> return ""
            ("" == code) -> return ""
        }

        return code
    }
}
